### PR TITLE
Oxidize two qubit basis decomposer

### LINKED ISSUE #12004: Port TwoQubitBasisDecomposition to Rust

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit is the second part of migrating the default 2q unitary
synthesis method to leverage parallel rust as described in #8774. The
Eventual goal is to be able to run unitary synthesis in parallel for all
the unitary matrices in a single call from the `UnitarySynthesis` pass.
The TwoQubitBasisDecomposer class is one of the default decomposers used
by the unitary synthesis plugin. After this we can build an interface
that will run the decomposition in parallel for a given decomposer.
    
This commit re-implements the TwoQubitBasisDecomposer class in rust. It
keeps the same algorithm from the previous python version but implements
it in rust. This builds off of #11946 and for the operation of the
decomposer class the TwoQubitWeylDecomposition class is used solely
through rust.
    
This commit depends on #11946 and will need to be rebased after #11946
is merged.
    
### Details and comments

Fixes #12004

TODO:

- [x] Wait for #11946 to merge and rebase
- [x] Fix last test failures (most fixing the error logic and a couple edge cases in the pulse optimal path)
- [x] Add release note

### CODE REVIEW

**Comment by mtreinish** [comment]
For some initial numbers I ran the same asv benchmarks that I did in https://github.com/Qiskit/qiskit/pull/11946#issuecomment-1989227610 to compare this PR against the current state of main (which doesn't have #11946 yet)

```
Benchmarks that have improved:

       before           after         ratio
     [230e91cb]       [627269be]
     <main>       <oxidize-two-qubit-basis-decomposer>
-     4.00±0.01ms      3.16±0.02ms     0.79  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'synthesis')
-      2.20±0.01s          1.68±0s     0.77  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3)
-      1.77±0.02s          1.33±0s     0.75  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)
-         1.39±0s          948±3ms     0.68  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0)
-       335±0.9ms        228±0.8ms     0.68  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2)
-         386±1ms        258±0.8ms     0.67  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3)
-     8.81±0.09ms      4.72±0.02ms     0.54  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'synthesis')
-         231±1ms        121±0.5ms     0.52  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0)
-     12.5±0.08ms      6.18±0.03ms     0.49  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')
-       208±0.5ms        101±0.2ms     0.49  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1)
-         825±6ms          390±2ms     0.47  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)
-      17.4±0.4ms       8.13±0.2ms     0.47  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'translator')
-       209±0.6ms       97.1±0.6ms     0.47  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0)
-       189±0.5ms       82.1±0.3ms     0.43  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1)
-        19.4±4ms       8.37±0.8ms     0.43  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'synthesis')
-      50.8±0.3ms      20.2±0.09ms     0.40  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'translator')
-       151±0.4ms       54.6±0.4ms     0.36  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'translator')
-         208±6ms         73.2±4ms     0.35  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'synthesis')
-       454±0.5ms        156±0.9ms     0.34  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'translator')
-      1.66±0.01s          569±2ms     0.34  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'translator')
-       931±0.6ms        319±0.4ms     0.34  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'translator')
-        54.9±2ms         18.2±2ms     0.33  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'synthesis')
-        864±30ms         282±20ms     0.33  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'synthesis')
-      2.10±0.01s         685±10ms     0.33  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'synthesis')
-      4.31±0.03s       1.39±0.02s     0.32  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'synthesis')

Benchmarks that have stayed the same:

       before           after         ratio
     [230e91cb]       [627269be]
     <main>       <oxidize-two-qubit-basis-decomposer>
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 100, 'lookahead')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 100, 'lookahead')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 100, 'lookahead')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 100, 'lookahead')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 100, 'lookahead')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 100, 'decay')
              n/a              n/a      n/a  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 100, 'lookahead')
      2.64±0.01ms      2.70±0.03ms     1.02  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'translator')
       3.11±0.02s       3.17±0.01s     1.02  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'decay')
          191±1ms          193±1ms     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'lookahead')
       21.4±0.2ms       21.6±0.1ms     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'lookahead')
        202±0.4ms          204±2ms     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'decay')
          182±1ms          183±1ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2)
        112±0.3ms        112±0.7ms     1.01  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1)
       1.87±0.01s          1.88±0s     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'lookahead')
        114±0.5ms        114±0.3ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1)
       23.1±0.2ms       23.2±0.1ms     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'decay')
          290±1ms          290±2ms     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'decay')
        184±0.8ms          184±2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2)
            11498            11498     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 10, 'decay')
            13134            13134     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 10, 'lookahead')
              557              557     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 10, 'decay')
              547              547     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 10, 'lookahead')
             5231             5231     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 100, 'decay')
             5663             5663     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 100, 'lookahead')
             2955             2955     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 10, 'decay')
             3945             3945     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 10, 'lookahead')
           217448           217448     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 10, 'decay')
           189590           189590     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 10, 'lookahead')
             4419             4419     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 10, 'decay')
             4351             4351     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 10, 'lookahead')
            44362            44362     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 100, 'decay')
            44711            44711     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 100, 'lookahead')
            42115            42115     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 10, 'decay')
            47894            47894     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 10, 'lookahead')
             2565             2565     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(0)
             1403             1403     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(1)
             1403             1403     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(2)
             1296             1296     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(3)
             2705             2705     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(0)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(1)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(2)
                7                7     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(3)
             2705             2705     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(0)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(1)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(2)
                7                7     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(3)
              323              323     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(0)
              336              336     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(1)
              336              336     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(2)
              272              272     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(3)
      53.5±0.07ms       53.3±0.4ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0)
        300±0.8ms          298±3ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'lookahead')
       51.4±0.1ms       50.6±0.4ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0)
       80.2±0.4ms       78.9±0.7ms     0.98  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3)
       82.3±0.3ms       79.8±0.2ms     0.97  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by mtreinish** [comment]
This should be ready to go now. There might be some more performance tuning and optimization we can do to the code (please feel free to leave any suggestions inline) but the performance is looking really good in it's current form. I reran the benchmarks from [above](https://github.com/Qiskit/qiskit/pull/12010#issuecomment-1996721745) against current main (which includes #11946 now) and it resulted in: 

```
Benchmarks that have improved:

       before           after         ratio
     [00b0952c]       [4320a623]
     <main>       <oxidize-two-qubit-basis-decomposer>
-      1.50±0.03s       1.36±0.01s     0.91  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(2)
-      1.02±0.02s          902±6ms     0.88  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(0)
-         304±2ms          267±3ms     0.88  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(3)
-       154±0.3ms          121±2ms     0.78  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(0)
-       138±0.4ms          105±2ms     0.77  transpiler_levels.TranspilerLevelBenchmarks.time_schedule_qv_14_x_14(1)
-       129±0.1ms         97.7±2ms     0.76  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(0)
-         534±3ms          398±5ms     0.74  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(1)
-       120±0.9ms       84.5±0.3ms     0.70  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(1)
-      7.78±0.5ms       4.40±0.2ms     0.57  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'synthesis')
-        14.0±3ms         7.14±1ms     0.51  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'synthesis')
-      11.5±0.3ms       5.41±0.7ms     0.47  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(2, 'translator')
-      15.7±0.4ms       6.90±0.3ms     0.44  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(3, 'translator')
-      46.4±0.1ms       16.4±0.2ms     0.35  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'translator')
-        51.9±4ms       16.9±0.7ms     0.33  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(5, 'synthesis')
-       138±0.9ms       42.5±0.5ms     0.31  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'translator')
-        717±20ms          215±7ms     0.30  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'synthesis')
-      1.49±0.01s          445±4ms     0.30  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'translator')
-         415±3ms          123±2ms     0.30  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(14, 'translator')
-         846±2ms          246±2ms     0.29  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'translator')
-        190±10ms         54.8±3ms     0.29  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(8, 'synthesis')
-      3.60±0.09s          945±7ms     0.26  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(27, 'synthesis')
-      1.84±0.03s          474±8ms     0.26  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(20, 'synthesis')

Benchmarks that have stayed the same:

       before           after         ratio
     [00b0952c]       [4320a623]
     <main>       <oxidize-two-qubit-basis-decomposer>
         80.9±1ms         82.2±2ms     1.02  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(3)
       3.19±0.01s       3.22±0.01s     1.01  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'decay')
       1.93±0.01s       1.94±0.01s     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(1081, 10, 'lookahead')
        115±0.5ms        115±0.9ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(1)
       55.0±0.1ms       55.1±0.3ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(0)
         52.6±1ms      52.7±0.09ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(0)
          181±2ms        181±0.2ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(2)
            11498            11498     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 10, 'decay')
            13134            13134     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(1081, 10, 'lookahead')
              557              557     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 10, 'decay')
              547              547     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 10, 'lookahead')
             5231             5231     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 100, 'decay')
             5663             5663     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(115, 100, 'lookahead')
             2955             2955     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 10, 'decay')
             3945             3945     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_depth_sabre_swap(409, 10, 'lookahead')
           217448           217448     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 10, 'decay')
           189590           189590     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(1081, 10, 'lookahead')
             4419             4419     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 10, 'decay')
             4351             4351     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 10, 'lookahead')
            44362            44362     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 100, 'decay')
            44711            44711     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(115, 100, 'lookahead')
            42115            42115     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 10, 'decay')
            47894            47894     1.00  quantum_volume.LargeQuantumVolumeMappingTrackBench.track_size_sabre_swap(409, 10, 'lookahead')
             2565             2565     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(0)
             1403             1403     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(1)
             1403             1403     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(2)
             1296             1296     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_quantum_volume_transpile_50_x_20(3)
             2705             2705     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(0)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(1)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(2)
                7                7     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm(3)
             2705             2705     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(0)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(1)
             2005             2005     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(2)
                7                7     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_from_large_qasm_backend_with_prop(3)
              323              323     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(0)
              336              336     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(1)
              336              336     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(2)
              272              272     1.00  transpiler_levels.TranspilerLevelBenchmarks.track_depth_transpile_qv_14_x_14(3)
        114±0.5ms        114±0.8ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(1)
       82.6±0.5ms       82.5±0.4ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm_backend_with_prop(3)
          181±2ms        181±0.5ms     1.00  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_from_large_qasm(2)
        296±0.8ms          296±1ms     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'decay')
       2.61±0.1ms      2.60±0.05ms     1.00  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'translator')
       22.0±0.2ms       22.0±0.1ms     1.00  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'lookahead')
          200±1ms          198±1ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'lookahead')
          210±2ms          207±1ms     0.99  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 100, 'decay')
          316±4ms          307±2ms     0.97  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(409, 10, 'lookahead')
       3.28±0.1ms      3.10±0.03ms     0.95  quantum_volume.QuantumVolumeBenchmark.time_ibmq_backend_transpile(1, 'synthesis')
       1.84±0.01s       1.71±0.01s     0.93  transpiler_levels.TranspilerLevelBenchmarks.time_quantum_volume_transpile_50_x_20(3)
       25.3±0.8ms       23.5±0.1ms     0.93  quantum_volume.LargeQuantumVolumeMappingTimeBench.time_sabre_swap(115, 10, 'decay')
          265±2ms          241±1ms     0.91  transpiler_levels.TranspilerLevelBenchmarks.time_transpile_qv_14_x_14(2)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: qiskit/transpiler/passes/synthesis/unitary_synthesis.py [hunk 1]
```diff
@@ -811,7 +811,7 @@ def is_controlled(gate):
                     pi2_decomposer = TwoQubitBasisDecomposer(
                         pi2_basis,
                         euler_basis=basis_1q,
-                        basis_fidelity=basis_2q_fidelity,
+                        #                        basis_fidelity=basis_2q_fidelity,
```
Comment (mtreinish) [inline qiskit/transpiler/passes/synthesis/unitary_synthesis.py]: Oops I forgot about this, I need to come up with a fix for this. `basis_2q_fidelity` is a dictionary here and that is failing when it gets passed to the rust code because the type is invalid it needs to be `float`. I commented this out to workaround it in the short term during development until I could debug what the intent was with the dictionary being passed in is.

Comment (mtreinish) [reply qiskit/transpiler/passes/synthesis/unitary_synthesis.py]: Fixed by: https://github.com/Qiskit/qiskit/pull/12010/commits/99eaa7b55482942ff5025d81a6f5682fbe72de90

File: crates/accelerate/src/two_qubit_decompose.rs [hunk 2]
```diff
@@ -1190,12 +1281,768 @@ impl TwoQubitGateSequence {
...
+            unitary,
+            &target_1q_basis_list,
+            qubit as usize,
+            None,
+            true,
+            None,
+        );
+        if let Some(sequence) = sequence {
+            *global_phase += sequence.global_phase;
+            for gate in sequence.gates {
```
Comment (kevinhartman) [inline crates/accelerate/src/two_qubit_decompose.rs:1599]: I'm not sure how large the sequence would be, but is there benefit to having `unitary_to_gate_sequence_inner` write directly to this instead of it returning a new vec?

File: crates/accelerate/src/two_qubit_decompose.rs [hunk 3]
```diff
@@ -1190,12 +1281,770 @@ impl TwoQubitGateSequence {
...
+        let mut gates = Vec::new();
+        let mut global_phase = target_decomposed.global_phase;
+        global_phase -= best_nbasis as f64 * self.basis_decomposer.global_phase;
+        if best_nbasis == 2 {
+            global_phase += PI;
+        }
+        for i in 0..best_nbasis as usize {
+            if let Some(euler_decomp) = &euler_decompositions[2 * i] {
+                for gate in &euler_decomp.gates {
+                    gates.push((gate.0.clone(), gate.1.clone(), smallvec![0]));
```
Comment (kevinhartman) [inline crates/accelerate/src/two_qubit_decompose.rs:2013]: Perhaps it makes sense to reserve the size of the decomp before pushing in these loops?

**Review by kevinhartman — COMMENTED** [review]
This looks mostly good to me, at least from a Rust perspective.

The performance improvements are fantastic! 😄

File: crates/accelerate/src/two_qubit_decompose.rs (hunk 2 above)
Comment (mtreinish) [reply crates/accelerate/src/two_qubit_decompose.rs:1599]: The sequence size depends on the selected euler basis, it's between 1 and 5 for all the ones defined in there. You can see where this is done in: https://github.com/Qiskit/qiskit/blob/main/crates/accelerate/src/euler_one_qubit_decomposer.rs (the `circuit_*` functions). The reason we can't write to it directly though is because the elements are different. The vec inside `OneQubitGateSequence` is a different type than what's in `TwoQubitGateSequence`. The two qubit variant has to track which qubit(s) the gate is on, while in the 1 qubit case we don't need to track that (because there is only one qubit).

File: crates/accelerate/src/two_qubit_decompose.rs (hunk 3 above)
Comment (mtreinish) [reply crates/accelerate/src/two_qubit_decompose.rs:2013]: I think we can do that. We might need a bit of logic to try and accurately precompute the vector lengths based on the euler basis accurately. Although maybe we can just over-allocate and assume the euler sequences are always the longest possible. It's only going to be a couple of entries so the over-allocation would be lost in the noise (and it's only temporary until we create the circuits. What do you think?

Comment (mtreinish) [reply crates/accelerate/src/two_qubit_decompose.rs:2013]: I did the worst case pre-allocation in: https://github.com/Qiskit/qiskit/pull/12010/commits/84c6dd42d5de09fb2ea6a82e598b91efe81db6e9 and left a detailed comment there that explains the worst case allocation is over allocating slightly and explaining what the at over allocation was. In the worst case it's only 16 extra elements, but since all 3 fields in the tuple are heap allocated we're only storing a pointer (and whatever else gets kept in a box) plus a few integers for each small vec in the  over allocated entries. It's also only a worst case over allocation by 16 elements which isn't that big a deal since we're not keeping the `TwoQubitGateSequence` objects around longer term (it also will only come up if the euler basis is `U` or `U3`, which aren't that common for real hardware).

Comment (kevinhartman) [reply crates/accelerate/src/two_qubit_decompose.rs:2013]: Seems fine to me!

**Review by kevinhartman — APPROVED** [review]

**Comment by mtreinish** [comment]
As keen as I am to see this merge I've pulled it from the merge queue just to give @levbishop a chance to review it too.

**Review by levbishop — APPROVED** [review]
This is generally fine as a translation of the python.

I do worry that it further locks in some possibly non-optimal choices from the python. In particular forcing an explicit 1q decomposition at 2q decomposition level is probably not the best overall strategy for the future (since it will almost certainly be immediately converted back to a unitary matrix and collected, resynthesized by the 1q optimizations)

Tracking the `num_basis_gates()` flow through the code is a bit of a maze. Python `num_basis_gates` calls rust `num_basis_gates` callls `_num_basis_gates` calls `__num_basis_gates`

I'm interested to try removing the layer of wrapping of the rust object inside a python object, but i think we can explore that in future PRs without breaking any API or anything.

I do very much like the perf delta from this work!

**Comment by mtreinish** [comment]
Yeah I think there is a round of clean up we'll want to do to simplify things. As you say I think that'll be fairly straightforward  once this merges without any api implications. Porting the algorithm was large and kind of unwieldy that I wasn't super concerned with the exact interface thinking we could iterate on it later, especially because the rust code is explicitly private.
